### ISSUE TITLE
CloudSpeechAPIService is not defined

### ISSUE THREAD

**Issue by maciz84** [GoogleCloudPlatform/google-cloud-dotnet#979]
Hi guys,

I am getting CloudSpeechAPIService is not defined in the below namespaces:

Imports Google.Apis.Auth.OAuth2
Imports Google.Cloud.Speech.V1
Imports Google.Apis.Services

I am trying to use this to authenticate as per the samples?

**Comment by maciz84** [comment]
Using this code as per the sample

`        static public CloudSpeechAPIService CreateAuthorizedClient()
        {
            GoogleCredential credential =
                GoogleCredential.GetApplicationDefaultAsync().Result;
            // Inject the Cloud Storage scope if required.
            if (credential.IsCreateScopedRequired)
            {
                credential = credential.CreateScoped(new[]
                {
                    CloudSpeechAPIService.Scope.CloudPlatform
                });
            }
            return new CloudSpeechAPIService(new BaseClientService.Initializer()
            {
                HttpClientInitializer = credential,
                ApplicationName = "DotNet Google Cloud Platform Speech Sample",
            });
        }`

**Comment by jskeet** [comment]
Which sample are you using? I suspect you're using the samples for the wrong library - using the REST-based library. I think we may still be waiting on the `Google.Cloud.Speech.V1` full samples, but there are examples here: https://googlecloudplatform.github.io/google-cloud-dotnet/docs/Google.Cloud.Speech.V1/index.html

(You won't end up with `CloudSpeechAPIService` at all.)

**Comment by maciz84** [comment]
Hi @jskeet 

Thank you for the link. I used the same link already to get my code to work yesterday. Basically I thought I would need this in order to get the progress of my Longrunning operation as I read somewhere that I will keep getting 404 errors when trying to run OperationAsync functions?? It said that one needs to use the same API key for the request in order to call https://speech.googleapis.com/v1beta1/operations/YOUR_OPERATION_NAME?key=YOUR_API_KEY

Basically I would like to get the progress of the operation and I don't know how to do this?

**Comment by jskeet** [comment]
No, you don't need a different library at all. Firstly, note that the link there is with v1beta1 - which is now out of date. Next, to use long running operations you just call the `LongRunningRecognize` method on the `SpeechClient`. That will return an `Operation<LongRunningRecognizeResponse>` which you can poll until it has completed, or find and store the operation name to poll later.

**Comment by maciz84** [comment]
Ok excellent,

So now I have this code which works as expected:

`dynamic audiofile = "gs://myBucket/test2.wav";

dynamic speech = SpeechClient.Create();
dynamic longOperation = speech.LongRunningRecognizeAsync(new RecognitionConfig {
	Encoding = RecognitionConfig.Types.AudioEncoding.Linear16,
	SampleRateHertz = 8000,
	LanguageCode = LanguageCodes.English.UnitedKingdom
}, RecognitionAudio.FromStorageUri(audiofile));

longOperation = longOperation.PollUntilCompletedAsync();

dynamic response = longOperation.Result;

foreach (object result in response.Results) {
	foreach (object alternative in result.Alternatives) {
		//txtOperation.Text = longOperation.RpcMessage.Name
		//txtReturn.Text = txtReturn.Text & alternative.Transcript
		MyObservable.Add(new TranscriptionResult {
			Transcript = alternative.Transcript,
			Confidence = alternative.Confidence
		});

	}
}`

This works well. I know I am asking you a lot of questions probably need to be in a new thread, however I would like to get the progress as per the sample json below:

`{
  "name": "5543203840552489181",
  "metadata": {
    "@type": "type.googleapis.com/google.cloud.speech.v1beta1.AsyncRecognizeMetadata",
    **"progressPercent": 100,**
    "startTime": "2017-02-01T22:11:26.469835Z",
    "lastUpdateTime": "2017-02-01T22:11:29.455780Z"
  },
  "done": true,
  "response": {
    "@type": "type.googleapis.com/google.cloud.speech.v1beta1.AsyncRecognizeResponse",
    "results": [
      {
        "alternatives": [
          {
            "transcript": "how old is the Brooklyn Bridge",
            "confidence": 0.98267895
          }
        ]
      }
    ]
  }
}`

Whilst it is processing...is this possible?

**Comment by jskeet** [comment]
Firstly, I'd *strongly* encourage you to stop using `dynamic` when you really don't need dynamic typing. All you're doing is stopping the compiler (and Intellisense) from helping you.

As for the progress metadata - this is something I haven't seen before in long running operations... it's not available now, but I'll add it as a feature request. It looks like it'll require changes within GAX as well to expose...

**Comment by jskeet** [comment]
(I'll close this issue as the originally reported issue is now resolved.)

**Comment by maciz84** [comment]
Thanks @jskeet yeah the dynamic thing was a conversion from a vb.net application that I am fixing for my colleague so I am using a conversion tool lol.

Cheers for the help again

**Comment by jskeet** [comment]
One extra note - while we don't expose it in any particularly useful way at the moment, you *could* get at the metadata from an existing `Operation<T>` with:

```csharp
var metadata = operation.RpcMessage.Unpack<LongRunningRecognizeMetadata>();
```

... but I'd like to expose that more cleanly.

**Comment by maciz84** [comment]
Ok so it is possible using Unpack?

**Comment by maciz84** [comment]
Would it be safe for me to use it now until you guys have exposed it cleanly?

**Comment by jskeet** [comment]
It should be safe to use, but:

- You'll need to control the polling yourself rather than using `PollUntilCompleted`
- You should check whether it's null first. (Simplest approach is probably to use `operation.RpcMessage?.Metadata.Unpack<LongRunningRecognizeMetadata>();`)

**Comment by maciz84** [comment]
Got it, thanks so much again you have been really, really helpful and I look forward to hopefully see the new features implemented soon :):)
